app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt

```diff
@@ -6,86 +6,49 @@ import androidx.appcompat.app.AppCompatActivity
-        }
-        binding.buttonMinus.setOnClickListener {
-            expression += Minus
-            binding.textView.text = expression.toString()
-        }
-        binding.buttonMultiply.setOnClickListener {
-            expression += Multiply
-            binding.textView.text = expression.toString()
-        }
-        binding.buttonDivide.setOnClickListener {
-            expression += Divide
-            binding.textView.text = expression.toString()
-        }
-        binding.buttonDelete.setOnClickListener {
-            expression = expression.removeLast()
-            binding.textView.text = expression.toString()
-        }
-        binding.buttonEquals.setOnClickListener {
-            val result = calculator.calculate(expression.toString())
-            if (result == null) {
-                Toast.makeText(this, R.string.incomplete_expression, Toast.LENGTH_SHORT).show()
-                return@setOnClickListener
-            }
-            expression = Expression.EMPTY + result
-            binding.textView.text = result.toString()
+        binding.button0.setOnClickListener { presenter.addNumberToExpression(0) }
+        binding.button1.setOnClickListener { presenter.addNumberToExpression(1) }
+        binding.button2.setOnClickListener { presenter.addNumberToExpression(2) }
+        binding.button3.setOnClickListener { presenter.addNumberToExpression(3) }
+        binding.button4.setOnClickListener { presenter.addNumberToExpression(4) }
+        binding.button5.setOnClickListener { presenter.addNumberToExpression(5) }
+        binding.button6.setOnClickListener { presenter.addNumberToExpression(6) }
+        binding.button7.setOnClickListener { presenter.addNumberToExpression(7) }
+        binding.button8.setOnClickListener { presenter.addNumberToExpression(8) }
+        binding.button9.setOnClickListener { presenter.addNumberToExpression(9) }
+
+        binding.buttonPlus.setOnClickListener { presenter.addOperatorToExpression(Operator.Plus) }
+        binding.buttonMinus.setOnClickListener { presenter.addOperatorToExpression(Operator.Minus) }
+        binding.buttonMultiply.setOnClickListener { presenter.addOperatorToExpression(Operator.Multiply) }
+        binding.buttonDivide.setOnClickListener { presenter.addOperatorToExpression(Operator.Divide) }
+
+        binding.buttonDelete.setOnClickListener { presenter.removeLastToken() }
+        binding.buttonEquals.setOnClickListener { presenter.calculateCurrentExpression() }
+    }
+
+    override fun showExpression(expression: Expression) {
+        binding.textView.text = expression.toString()
+    }
+
+    override fun showResult(result: Int) {
+        binding.textView.text = result.toString()
+    }
+
+    override fun showErrorMessage(exception: Exception) {
+        when (exception) {
+            is IncompleteExpressionException -> Toast.makeText(this, R.string.incomplete_expression, Toast.LENGTH_SHORT).show()
+            else -> {}
         }
```

Comment:
단순히 하나의 조건일 경우에는 when보다는 if문을 사용하는 것을 선호해요.

> Prefer using if for binary conditions instead of when

ref. https://kotlinlang.org/docs/coding-conventions.html#if-versus-when